### PR TITLE
executor,server: re-implement the `kill` statement by checking the `Next()` function

### PR DESCRIPTION
**Pull request by tiancaiamao**
### What problem does this PR solve? 

Re-implement the `kill` statement in a better way.

In https://github.com/pingcap/tidb/pull/9844, the implementation of `kill` is changed to use `executor.Close`. Some goroutine may be calling `executor.Next` while another one is calling the `executor.Close`. It's not thread-safe, and panic sometimes (you can imagine that one of them is using `e.result` while another one set `e.result` to `nil`).

### What is changed and how it works?

In this commit, I introduce a `Next()` function that wraps the `e.Next()` function, and `ctx.Done()` is checked in the new `Next()` function.
In this way, we can make sure `kill` signal could be checked as long as `Next()` is called, so we can stop an executor eventually.

### Check List 

Tests 

It's not easy to write unit tests for the `kill` statement. I tried to make a 'black hole' table (which generates fake data forever when it's read) to support long-running transactions for `kill`, but it involves many more changes and those changes are irrelevant. So I decide to leave tests along until I find some better ways.

Code changes

 - Has exported function/method change

Side effects

 - Possible performance regression

Related changes

 - Need to cherry-pick to the release branch

### CODE REVIEW

**Comment by tiancaiamao** [comment]
PTAL @lysu @qw4990 @XuHuaiyu

File: server/server.go [hunk 1]
```diff
@@ -530,15 +530,8 @@ func (s *Server) Kill(connectionID uint64, query bool) {
 
 func killConn(conn *clientConn) {
 	conn.mu.RLock()
-	resultSets := conn.mu.resultSets
 	cancelFunc := conn.mu.cancelFunc
 	conn.mu.RUnlock()
-	for _, resultSet := range resultSets {
-		// resultSet.Close() is reentrant so it's safe to kill a same connID multiple times
-		if err := resultSet.Close(); err != nil {
```
Comment (crazycs520) [inline server/server.go]: As we won't call `Close` now, then, how to make sure the executor stop running and release resource?

Comment (crazycs520) [reply server/server.go]: Check `context.Contex` in every `executor.Next`?

File: executor/executor.go [hunk 2]
```diff
@@ -175,6 +175,17 @@ type Executor interface {
 	newFirstChunk() *chunk.Chunk
 }
 
+// Next is a wrapper function on e.Next(), it handles some common codes.
+func Next(ctx context.Context, e Executor, req *chunk.RecordBatch) error {
```
Comment (lonng) [inline executor/executor.go]: Do we need to export this function?

File: executor/executor.go [hunk 3]
```diff
@@ -175,6 +175,17 @@ type Executor interface {
...
+	select {
+	case <-ctx.Done():
+		req.Reset()
+		return errors.Trace(ctx.Err())
```
Comment (zz-jason) [inline executor/executor.go]: no need to trace

File: executor/executor.go [hunk 4]
```diff
@@ -175,6 +175,17 @@ type Executor interface {
 	newFirstChunk() *chunk.Chunk
 }
 
+// Next is a wrapper function on e.Next(), it handles some common codes.
+func Next(ctx context.Context, e Executor, req *chunk.RecordBatch) error {
+	select {
```
Comment (zz-jason) [inline executor/executor.go]: what's the performance regression for large queries which require to call a huge amount of Next() function?

File: server/server.go (hunk 1 above)
Comment (tiancaiamao) [reply server/server.go]: It works like this:
tidb receive `kill` input =>
session manager call the cancel function =>
the goroutine of the killed session run to the `Next()` function =>
check ctx.Done() in `Next()` and return error =>
the caller find `Next()` return error, it call `Close()` =>
finish current transaction

File: executor/executor.go (hunk 2 above)
Comment (tiancaiamao) [reply executor/executor.go]: It's better to export this function.
As `e.Next()` is replaced by `Next()` now, calling `e.Next()` doesn't reflect the truth.
We may call `execute.Next(e)` outside the execute package, for example, writing tests.

File: executor/executor.go (hunk 3 above)
Comment (tiancaiamao) [reply executor/executor.go]: This is not a defined error in TiDB, I'm afraid without trace there won't be an error stack.
Or, should we define an error for it?

**Comment by lonng** [comment]
Can we use an atomic variable to determine whether the `Executor` was been canceled? I think it's more effective than check `context.Done()` in every iteration.

**Comment by tiancaiamao** [comment]
> Can we use an atomic variable to determine whether the Executor was been canceled? I think it's more effective than check context.Done() in every iteration.

Use many different mechanisms to implement the same thing makes the code less maintainable.
And another drawback of checking atomic variable is that you can't handle block waiting conditions:

```
select {
    case <-wait:
    for atomic is false ? {
    }
}
```
 @lonng

**Comment by tiancaiamao** [comment]
> 1. context should be passed from the root executor to the deep leaf executor in the tree.

No problem... passing context is the golang idiom

> 2. context.Done() should be checked in every background goroutine for every parallel executor. And once the upper layer calls Cancel(), all the background goroutines should stop.

I don't mean to check **every** background goroutine any more. It's impractical as we have proved: difficult to maintain and easy to cause goroutine leak.
The parallel background goroutine will not check the `ctx.Done` any more, instead, it handles those errors `Next()` returned.
In this way, we can guarantee the cancel of the execution within two `Next` call.
Maybe we still need some check of `ctx.Done` in the parallel executor goroutine (if the `Next` itself involves too many computing), but I don't want to make the `ctx.Done` check diffuse.

> 3. Close() should be safe to call after the upper layer called Cancel().

Cancel will not cancel the background goroutine now ... it just makes `Next` return error.
So the release of resources should be safe.

**Comment by lonng** [comment]
@tiancaiamao I mean to set an atomic variable to 1 when killing an executor and check the atomic variable while every iteration.

**Comment by tiancaiamao** [comment]
I know an atomic variable is more effective, but it's not the only factor we should take into consideration @lonng

**Comment by qw4990** [comment]
Can we guarantee that after canceling an Executor, the `Close()` function of this Executor must be called by some goroutine, otherwise resources leak would happen?

**Comment by tiancaiamao** [comment]
`Close` will be called in the session's goroutine, this is the same as before (I mean before PR https://github.com/pingcap/tidb/pull/9844) 
Calling `Close` even when `executor.Next` meets error, is a prerequisite.
@qw4990

**Comment by lysu** [comment]
there are maybe many `Next()` calls need to check ctx.Done...

for example:
```
   limit - unparallel project - tablescan
```
check `ctx.Done` seems be not well- -... check in limit then go to project quickly but need check again...

and high frequent select `ctx.Done` make pressure to schedule, maybe just check a shared variable in wrapped `Next` method is better?

maybe need a benchmark- -?

**Comment by tiancaiamao** [comment]
Now I change the PR to use an atomic flag @zz-jason @lysu @lonng @qw4990

**Pull request merged**